File: sdk/python/haltchain/langgraph.py
"""LangGraph integration — drop-in HaltChain safety for any LangGraph workflow.

Install extras:  pip install haltchain[langgraph]

Usage (3 lines to add HaltChain)::

    from haltchain.langgraph import HaltChainGuard

    guard = HaltChainGuard(agent_id="my-agent", api_key="key")
    graph = guard.wrap(graph_builder.compile())

The guard automatically intercepts tool calls and state transitions,
validating each against HaltChain policy before execution proceeds.

For finer control, use the guard as a node or interrupt::

    # As a validation node in your graph
    builder.add_node("haltchain", guard.as_node())
    builder.add_edge("agent", "haltchain")
    builder.add_edge("haltchain", "tools")

    # As a conditional gate
    builder.add_conditional_edges(
        "agent",
        guard.gate(then="tools", deny="end"),
    )
"""

from __future__ import annotations

import functools
import json
from typing import Any, Callable, Dict, Optional, Sequence, TypeVar

from .async_client import AsyncHaltChainClient
from .client import HaltChainClient
from .exceptions import CircuitBreakerError, PolicyViolationError

try:
    from langgraph.graph import StateGraph
    from langgraph.graph.graph import CompiledGraph
    from langgraph.prebuilt import ToolNode

    _LANGGRAPH_AVAILABLE = True
except ImportError:
    _LANGGRAPH_AVAILABLE = False


class HaltChainGuard:
    """Drop-in safety layer for LangGraph workflows.

    Parameters
    ----------
    agent_id : str
        Agent identifier registered with HaltChain.
    api_key : str
        HaltChain API key.
    base_url : str
        HaltChain validator URL.
    tool_action_map : dict, optional
        Maps tool names to callables that produce HaltChain action dicts
        from tool input. If omitted, ``{"type": tool_name, **input}`` is used.
    block_on_deny : bool
        If True (default), denied actions raise PolicyViolationError which
        triggers LangGraph error handling. If False, injects a deny message
        into the state instead.
    session_key : str
        State key used to store/read the session id (default: "session_id").
    """

    def __init__(
        self,
        *,
        agent_id: str,
        api_key: str,
        base_url: str = HaltChainClient.DEFAULT_BASE,
        tool_action_map: Optional[Dict[str, Callable[[dict], dict]]] = None,
        block_on_deny: bool = True,
        session_key: str = "session_id",
        **client_kwargs: Any,
    ) -> None:
        if not _LANGGRAPH_AVAILABLE:
            raise ImportError(
                "langgraph is required: pip install haltchain[langgraph]"
            )
        self._client = HaltChainClient(
            agent_id=agent_id,
            api_key=api_key,
            base_url=base_url,
            **client_kwargs,
        )
        self._async_client: Optional[AsyncHaltChainClient] = None
        self._agent_id = agent_id
        self._api_key = api_key
        self._base_url = base_url
        self._client_kwargs = client_kwargs
        self._tool_map = tool_action_map or {}
        self._block_on_deny = block_on_deny
        self._session_key = session_key

    def _get_async_client(self) -> AsyncHaltChainClient:
        if self._async_client is None:
            self._async_client = AsyncHaltChainClient(
                agent_id=self._agent_id,
                api_key=self._api_key,
                base_url=self._base_url,
                **self._client_kwargs,
            )
        return self._async_client

    # ── Public API ────────────────────────────────────────────────────────

    def wrap(self, compiled_graph: "CompiledGraph") -> "CompiledGraph":
        """Wrap a compiled LangGraph graph with HaltChain validation.

        Monkey-patches the graph's tool node (if present) to validate
        every tool call before execution. Returns the same graph object
        for chaining.

        Usage::

            graph = guard.wrap(builder.compile())
        """
        # Find and wrap ToolNode instances in the graph
        for node_name, node_fn in list(compiled_graph.nodes.items()):
            if isinstance(node_fn, ToolNode):
                compiled_graph.nodes[node_name] = self._wrap_tool_node(node_fn)
            elif hasattr(node_fn, "__wrapped_haltchain__"):
                continue  # Already wrapped
        return compiled_graph

    def as_node(self) -> Callable:
        """Return a LangGraph node function that validates the last message.

        Insert between the agent and tools node::

            builder.add_node("haltchain", guard.as_node())
        """

        async def _validate_node(state: dict) -> dict:
            messages = state.get("messages", [])
            if not messages:
                return state

            last_msg = messages[-1]
            tool_calls = getattr(last_msg, "tool_calls", None)
            if not tool_calls:
                return state

            session_id = state.get(self._session_key)

            for tc in tool_calls:
                tool_name = tc.get("name", tc.get("function", {}).get("name", "unknown"))
                tool_input = tc.get("args", tc.get("function", {}).get("arguments", {}))
                if isinstance(tool_input, str):
                    try:
                        tool_input = json.loads(tool_input)
                    except (json.JSONDecodeError, TypeError):
                        tool_input = {"input": tool_input}

                action = self._build_action(tool_name, tool_input)
                metadata = {
                    "framework": "langgraph",
                    "tool_name": tool_name,
                    "node_type": "validation_gate",
                }

                client = self._get_async_client()
                result = await client.check(action, metadata=metadata, session_id=session_id)
                decision = result.get("decision", "ALLOW")

                if decision in ("DENY", "CIRCUIT_BREAK"):
                    if self._block_on_deny:
                        raise PolicyViolationError(
                            decision=decision,
                            reason=result.get("reason", "Policy denied"),
                            policy=result.get("policy", ""),
                        )
                    # Non-blocking: return a message indicating denial
                    return {
                        "messages": [
                            {
                                "role": "system",
                                "content": f"[HaltChain DENIED] {result.get('reason', 'Policy violation')}",
                            }
                        ]
                    }

            return state

        _validate_node.__wrapped_haltchain__ = True  # type: ignore[attr-defined]
        return _validate_node

    def gate(
        self,
        *,
        then: str = "tools",
        deny: str = "__end__",
    ) -> Callable:
        """Return a conditional edge function for LangGraph.

        Routes to ``then`` on ALLOW, ``deny`` on DENY/CIRCUIT_BREAK::

            builder.add_conditional_edges("agent", guard.gate(then="tools", deny="end"))
        """

        def _route(state: dict) -> str:
            messages = state.get("messages", [])
            if not messages:
                return then

            last_msg = messages[-1]
            tool_calls = getattr(last_msg, "tool_calls", None)
            if not tool_calls:
                return then

            session_id = state.get(self._session_key)

            for tc in tool_calls:
                tool_name = tc.get("name", tc.get("function", {}).get("name", "unknown"))
                tool_input = tc.get("args", {})
                if isinstance(tool_input, str):
                    try:
                        tool_input = json.loads(tool_input)
                    except (json.JSONDecodeError, TypeError):
                        tool_input = {"input": tool_input}

                action = self._build_action(tool_name, tool_input)
                metadata = {"framework": "langgraph", "tool_name": tool_name}

                try:
                    result = self._client.check(
                        action, metadata=metadata, session_id=session_id,
                    )
                    if result.get("decision") in ("DENY", "CIRCUIT_BREAK"):
                        return deny
                except (PolicyViolationError, CircuitBreakerError):
                    return deny

            return then

        return _route

    # ── Internal helpers ──────────────────────────────────────────────────

    def _build_action(self, tool_name: str, tool_input: dict) -> dict:
        if tool_name in self._tool_map:
            return self._tool_map[tool_name](tool_input)
        return {"type": tool_name, **tool_input}

    def _wrap_tool_node(self, tool_node: "ToolNode") -> Callable:
        """Wrap a ToolNode's invoke to pre-validate each tool call."""
        original_invoke = tool_node.invoke

        @functools.wraps(original_invoke)
        def guarded_invoke(state: dict, config: Any = None) -> Any:
            messages = state.get("messages", [])
            if messages:
                last_msg = messages[-1]
                tool_calls = getattr(last_msg, "tool_calls", None)
                if tool_calls:
                    session_id = state.get(self._session_key)
                    for tc in tool_calls:
                        tool_name = tc.get("name", "unknown")
                        tool_input = tc.get("args", {})
                        action = self._build_action(tool_name, tool_input)
                        metadata = {"framework": "langgraph", "tool_name": tool_name}

                        result = self._client.check(
                            action, metadata=metadata, session_id=session_id,
                        )
                        decision = result.get("decision", "ALLOW")
                        if decision in ("DENY", "CIRCUIT_BREAK"):
                            raise PolicyViolationError(
                                decision=decision,
                                reason=result.get("reason", "Policy denied"),
                                policy=result.get("policy", ""),
                            )

            return original_invoke(state, config)

        guarded_invoke.__wrapped_haltchain__ = True  # type: ignore[attr-defined]
        return guarded_invoke

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> "HaltChainGuard":
        return self

    def __exit__(self, *_: Any) -> None:
        self.close()

File: sdk/typescript/src/langgraph.ts
/**
 * LangGraph integration — drop-in HaltChain safety for LangGraph.js workflows.
 *
 * Usage (3 lines to add HaltChain):
 *
 * ```ts
 * import { HaltCheckpointer } from "@haltchain/sdk/langgraph";
 *
 * const checkpointer = new HaltCheckpointer({ agentId: "my-agent", apiKey: "key" });
 * const safeGraph = checkpointer.wrap(compiledGraph);
 * ```
 *
 * For finer control, use as a checkpoint boundary interceptor:
 *
 * ```ts
 * builder.addNode("haltchain_checkpoint", checkpointer.asCheckpointNode());
 * ```
 *
 * Legacy `HaltChainGuard` is kept for backward compatibility.
 */

import { HaltChainClient } from "./client.js";
import type { HaltChainClientOptions, ValidationResponse } from "./types.js";

export interface LangGraphGuardOptions extends HaltChainClientOptions {
  /** Maps tool names to functions that build HaltChain action objects from tool input. */
  toolActionMap?: Record<string, (input: Record<string, unknown>) => Record<string, unknown>>;
  /** If true (default), denied actions throw. If false, injects a deny message. */
  blockOnDeny?: boolean;
  /** State key for session ID propagation. */
  sessionKey?: string;
}

/**
 * Drop-in safety layer for LangGraph.js workflows.
 *
 * Intercepts tool calls and state transitions, validating each against
 * HaltChain policy before execution proceeds.
 */
export class HaltChainGuard {
  private readonly client: HaltChainClient;
  private readonly toolMap: Record<string, (input: Record<string, unknown>) => Record<string, unknown>>;
  private readonly blockOnDeny: boolean;
  private readonly sessionKey: string;

  constructor(opts: LangGraphGuardOptions) {
    this.client = new HaltChainClient(opts);
    this.toolMap = opts.toolActionMap ?? {};
    this.blockOnDeny = opts.blockOnDeny ?? true;
    this.sessionKey = opts.sessionKey ?? "session_id";
  }

  /**
   * Validate an action and return the validation response.
   * Throws if denied and blockOnDeny is true.
   */
  async validate(
    action: Record<string, unknown>,
    sessionId?: string,
  ): Promise<ValidationResponse> {
    const result = await this.client.check(action, { sessionId });
    if (this.blockOnDeny && (result.decision === "DENY" || result.decision === "CIRCUIT_BREAK")) {
      throw new Error(`HaltChain blocked: ${result.reason ?? result.decision}`);
    }
    return result;
  }

  /**
   * Build an action dict from a tool call. Uses toolActionMap if available,
   * otherwise creates a generic `{ type: toolName, ...input }` action.
   */
  private buildAction(toolName: string, input: Record<string, unknown>): Record<string, unknown> {
    const mapper = this.toolMap[toolName];
    if (mapper) return mapper(input);
    return { type: toolName, ...input };
  }

  /**
   * Returns a node function compatible with LangGraph's `addNode()`.
   *
   * The node validates the last tool call in state.messages and either
   * passes through or throws/injects a deny message.
   */
  asNode(): (state: Record<string, unknown>) => Promise<Record<string, unknown>> {
    return async (state: Record<string, unknown>) => {
      const messages = state.messages as Array<Record<string, unknown>> | undefined;
      if (!messages?.length) return state;

      const last = messages[messages.length - 1];
      const toolCalls = (last as Record<string, unknown>).tool_calls as
        | Array<{ name: string; args: Record<string, unknown> }>
        | undefined;

      if (!toolCalls?.length) return state;

      const sessionId = state[this.sessionKey] as string | undefined;

      for (const call of toolCalls) {
        const action = this.buildAction(call.name, call.args);
        await this.validate(action, sessionId);
      }

      return state;
    };
  }

  /**
   * Returns a conditional gate function for `addConditionalEdges()`.
   *
   * Routes to `then` if validation passes, or `deny` if blocked.
   */
  gate(opts: { then: string; deny: string }): (state: Record<string, unknown>) => Promise<string> {
    return async (state: Record<string, unknown>) => {
      try {
        const messages = state.messages as Array<Record<string, unknown>> | undefined;
        if (!messages?.length) return opts.then;

        const last = messages[messages.length - 1];
        const toolCalls = (last as Record<string, unknown>).tool_calls as
          | Array<{ name: string; args: Record<string, unknown> }>
          | undefined;

        if (!toolCalls?.length) return opts.then;

        const sessionId = state[this.sessionKey] as string | undefined;

        for (const call of toolCalls) {
          const action = this.buildAction(call.name, call.args);
          const result = await this.client.check(action, { sessionId });
          if (result.decision === "DENY" || result.decision === "CIRCUIT_BREAK") {
            return opts.deny;
          }
        }
        return opts.then;
      } catch {
        return opts.deny;
      }
    };
  }

  /**
   * Wrap a compiled graph's invoke/stream to auto-validate tool calls.
   * Returns a proxy that intercepts invoke() and stream() calls.
   */
  wrap<T extends { invoke: (...args: unknown[]) => Promise<unknown> }>(graph: T): T {
    const guard = this;
    const originalInvoke = graph.invoke.bind(graph);

    const wrappedInvoke = async (...args: unknown[]): Promise<unknown> => {
      const input = args[0] as Record<string, unknown> | undefined;
      if (input) {
        const sessionId = (input[guard.sessionKey] as string) ?? undefined;
        // Pre-validate the input action if it looks like one
        if (input.type || input.action_type) {
          await guard.validate(input as Record<string, unknown>, sessionId);
        }
      }
      return originalInvoke(...args);
    };

    return new Proxy(graph, {
      get(target, prop) {
        if (prop === "invoke") return wrappedInvoke;
        return Reflect.get(target, prop);
      },
    });
  }
}

// ── Structured diff helpers ───────────────────────────────────────────────────

/** Flatten an object to leaf `"key.path": value` pairs for diff computation. */
function flattenState(obj: Record<string, unknown>, prefix = ""): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(obj)) {
    const key = prefix ? `${prefix}.${k}` : k;
    if (v !== null && typeof v === "object" && !Array.isArray(v)) {
      Object.assign(out, flattenState(v as Record<string, unknown>, key));
    } else {
      out[key] = v;
    }
  }
  return out;
}

/** Returns keys that changed between two flattened state snapshots. */
function stateDiff(
  before: Record<string, unknown>,
  after: Record<string, unknown>,
): Array<{ key: string; before: unknown; after: unknown }> {
  const keys = new Set([...Object.keys(before), ...Object.keys(after)]);
  const changes: Array<{ key: string; before: unknown; after: unknown }> = [];
  for (const key of keys) {
    if (JSON.stringify(before[key]) !== JSON.stringify(after[key])) {
      changes.push({ key, before: before[key], after: after[key] });
    }
  }
  return changes;
}

// ── HaltCheckpointer ─────────────────────────────────────────────────────────

export interface CheckpointerOptions extends HaltChainClientOptions {
  /**
   * State key holding the trace ID for cross-agent audit propagation.
   * Default: "haltchain_trace_id".
   */
  traceIdKey?: string;
  /**
   * State key holding the session ID for goal-drift tracking.
   * Default: "session_id".
   */
  sessionKey?: string;
  /**
   * Maximum number of diff entries to include in the metadata payload.
   * Prevents large state objects from bloating the request body.
   * Default: 50.
   */
  maxDiffEntries?: number;
  /** If true (default), throw on DENY/CIRCUIT_BREAK. */
  blockOnDeny?: boolean;
}

/**
 * HaltCheckpointer — intercepts LangGraph `StateGraph` transitions at the
 * checkpoint boundary, capturing pre/post state diffs automatically.
 *
 * This is the roadmap-compliant integration class (replaces HaltChainGuard
 * for new code).  It validates the **state transition** itself, not just
 * individual tool calls, enabling policy rules that inspect what actually
 * changed between graph nodes.
 *
 * Example:
 * ```ts
 * const checkpointer = new HaltCheckpointer({ agentId: "orchestrator", apiKey });
 * builder.addNode("haltchain_checkpoint", checkpointer.asCheckpointNode());
 * builder.addEdge("agent", "haltchain_checkpoint");
 * builder.addEdge("haltchain_checkpoint", "tools");
 * ```
 */
export class HaltCheckpointer {
  private readonly client: HaltChainClient;
  private readonly traceIdKey: string;
  private readonly sessionKey: string;
  private readonly maxDiffEntries: number;
  private readonly blockOnDeny: boolean;

  constructor(opts: CheckpointerOptions) {
    this.client = new HaltChainClient(opts);
    this.traceIdKey = opts.traceIdKey ?? "haltchain_trace_id";
    this.sessionKey = opts.sessionKey ?? "session_id";
    this.maxDiffEntries = opts.maxDiffEntries ?? 50;
    this.blockOnDeny = opts.blockOnDeny ?? true;
  }

  /**
   * Validate a state transition — the core checkpoint boundary check.
   *
   * @param stateBefore State snapshot before the node ran.
   * @param stateAfter  State snapshot after the node ran.
   * @param nodeName    Name of the completed node (used as action_type).
   */
  async checkTransition(
    stateBefore: Record<string, unknown>,
    stateAfter: Record<string, unknown>,
    nodeName: string,
  ): Promise<ValidationResponse> {
    const traceId = (stateAfter[this.traceIdKey] ?? stateBefore[this.traceIdKey]) as string | undefined;
    const sessionId = (stateAfter[this.sessionKey] ?? stateBefore[this.sessionKey]) as string | undefined;

    const flatBefore = flattenState(stateBefore);
    const flatAfter = flattenState(stateAfter);
    const diff = stateDiff(flatBefore, flatAfter).slice(0, this.maxDiffEntries);

    const action: Record<string, unknown> = {
      type: `checkpoint:${nodeName}`,
      // Surface tool calls in the diff so YAML rules can inspect them
      changed_keys: diff.map((d) => d.key),
      diff_count: diff.length,
      diff_sample: diff.slice(0, 10),
    };

    // Propagate trace ID as metadata for cross-agent audit correlation
    const metadata: Record<string, unknown> = {};
    if (traceId) metadata["haltchain_trace_id"] = traceId;

    const result = await this.client.check(action, { sessionId, context: metadata });

    if (
      this.blockOnDeny &&
      (result.decision === "DENY" || result.decision === "CIRCUIT_BREAK")
    ) {
      throw new Error(
        `HaltCheckpointer blocked at node '${nodeName}': ${result.reason ?? result.decision}`,
      );
    }

    return result;
  }

  /**
   * Returns a LangGraph node function that intercepts the checkpoint boundary.
   *
   * The node stores a pre-transition snapshot, lets the state pass through,
   * then validates the diff.  Place it after the agent node and before tools.
   */
  asCheckpointNode(
    nodeName = "haltchain_checkpoint",
  ): (state: Record<string, unknown>) => Promise<Record<string, unknown>> {
    return async (state: Record<string, unknown>) => {
      // The "before" snapshot is the state arriving at this node.
      // The "after" snapshot is the state produced by the previous node (same object here;
      // real diff is captured between this node's input and the *next* node's input).
      //
      // Since LangGraph passes immutable state objects, we treat the incoming state
      // as the post-transition snapshot and compare against the stored pre-snapshot.
      const preKey = `__haltchain_pre_${nodeName}`;
      const prevSnapshot = state[preKey] as Record<string, unknown> | undefined;
      const preSnapshot = prevSnapshot ?? {};

      await this.checkTransition(preSnapshot, state, nodeName);

      // Store current state as the baseline for the next checkpoint boundary.
      return { [preKey]: { ...state } };
    };
  }

  /**
   * Wrap a compiled StateGraph to automatically checkpoint every node transition.
   *
   * Injects a pre-invoke snapshot and validates the post-invoke state diff.
   */
  wrap<T extends { invoke: (...args: unknown[]) => Promise<unknown> }>(graph: T): T {
    const self = this;
    const originalInvoke = graph.invoke.bind(graph);

    const wrappedInvoke = async (...args: unknown[]): Promise<unknown> => {
      const input = (args[0] ?? {}) as Record<string, unknown>;
      const preSnapshot = flattenState(input);

      const output = (await originalInvoke(...args)) as Record<string, unknown>;

      // Validate the full graph transition (input → output)
      await self.checkTransition(preSnapshot, flattenState(output), "graph_invoke");

      return output;
    };

    return new Proxy(graph, {
      get(target, prop) {
        if (prop === "invoke") return wrappedInvoke;
        return Reflect.get(target, prop);
      },
    });
  }
}
